src/api_service.rs

```diff
@@ -222,27 +224,40 @@ pub fn run_upload_in_transaction(
         .get(&current_version)
         .map_err(|e| WaCustomError::BufIo(Arc::new(e)))?;
 
-    vecs.into_par_iter()
-        .map(|(id, vec)| {
+    let mut handles: Vec<thread::JoinHandle<Result<(), WaCustomError>>> = Vec::new();
+    let (tx, rx) = mpsc::channel();
+
+    for (id, vec) in vecs {
+        let tx = tx.clone();
+        let bufman = bufman.clone();
+        let dense_index = dense_index.clone();
+        let handle = thread::spawn(move || {
```

Comment:
Spawning a new thread for every element in the vec will definitely cause overhead. Should use something like [ThreadPool](https://docs.rs/rayon/latest/rayon/struct.ThreadPool.html)